src/Plugin/CheckImports.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | CheckImport plugin:
+--   This plugin checks if imported identifiers as unqualified
+--   exist and lists them.
+module Plugin.CheckImports
+  ( -- NOTE: The name "plugin" should be used to be called via GHC plugin mechanism.
+    plugin,
+  )
+where
+
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Char (isAlpha)
+import Data.Foldable (for_)
+import Data.IORef (readIORef)
+import Data.List (foldl', sort)
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Set (Set)
+import qualified Data.Set as S
+import GHC.Driver.Session (DynFlags, getDynFlags)
+import GHC.Plugins
+  ( CommandLineOption,
+    ModSummary,
+    Name,
+    Plugin (..),
+    defaultPlugin,
+    localiseName,
+    showSDoc,
+  )
+import GHC.Tc.Types (TcGblEnv (..), TcM)
+import GHC.Types.Name.Reader
+  ( GlobalRdrElt (..),
+    GreName (..),
+    ImpDeclSpec (..),
+    ImportSpec (..),
+  )
+import GHC.Unit.Module (ModuleName)
+import GHC.Utils.Outputable (Outputable (ppr))
+import Prelude hiding ((<>))
+
+plugin :: Plugin
+plugin =
+  defaultPlugin
+    { typeCheckResultAction = typecheckPlugin
+    }
+
+printPpr :: (Outputable a, MonadIO m) => DynFlags -> a -> m ()
+printPpr dflags = liftIO . putStrLn . showSDoc dflags . ppr
+
+formatName :: DynFlags -> Name -> String
+formatName dflags name =
+  let str = showSDoc dflags . ppr . localiseName $ name
+   in case str of
+        (x : _) ->
+          if isAlpha x
+            then str
+            else "(" ++ str ++ ")"
+        _ -> str
+
+formatImportedNames :: [String] -> String
+formatImportedNames names =
+  case fmap (++ ",\n") $ sort names of
+    l0 : ls ->
+      let l0' = "  ( " ++ l0
+          ls' = fmap ("    " ++) ls
+          footer = "  )"
+       in concat ([l0'] ++ ls' ++ [footer])
+    _ -> "  ()"
+
+mkModuleNameMap :: GlobalRdrElt -> [(ModuleName, Name)]
+mkModuleNameMap gre = do
+  spec <- gre_imp gre
+  case gre_name gre of
+    NormalGreName name -> do
+      let modName = is_mod $ is_decl spec
+      pure (modName, name)
+    _ -> []
+
+typecheckPlugin ::
+  [CommandLineOption] ->
+  ModSummary ->
+  TcGblEnv ->
+  TcM TcGblEnv
+typecheckPlugin _ modsummary tc = do
+  dflags <- getDynFlags
+  usedGREs :: [GlobalRdrElt] <-
+    liftIO $ readIORef (tcg_used_gres tc)
+  let moduleImportMap :: Map ModuleName (Set Name)
+      moduleImportMap =
+        foldl' (\(!m) (modu, name) -> M.insertWith S.union modu (S.singleton name) m) M.empty $
```

Comment:
I'm a little bit confused about why we're importing the lazy operations for `Map` and `Set` but then we're making sure to make the map strict when we get it. Am I missing something simple here? 😅

Reply:
oh. i mindlessly imported Data.Map, not Data.Map.Strict. Thanks for the suggestion.